uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-balance.report.component.spec.js

```diff
@@ -0,0 +1,213 @@
+          expect(ctrl.loading.report).toBe(true);
+        });
+
+        it('calls leaveBalanceReport.all with the page parameters', function () {
+          expect(leaveBalanceReport.all).toHaveBeenCalledWith(
+            { managed_by: loggedInContactId },
+            { page: page, size: defaultReportSize }
+          );
+        });
+
+        describe('when finishing loading a report`s page', function () {
+          var expected, count;
+
+          beforeEach(function () {
+            var absenceTypes = absenceTypeMock.all().values;
+            var balanceReport = reportMockData.all().values;
+            count = reportMockData.all().count;
+
+            // reversing absence_types to see if component sorts them in the
+            // proper order.
+            balanceReport.forEach(function (r) {
+              r.absence_types.reverse();
+            });
+
+            // sets the balance report in an expected manner.
+            // each record`s .absence_types should be sorted in the same order
+            // as the component`s .absenceTypes array.
+            expected = balanceReport.map(function (record) {
+              record = Object.assign({}, record);
+
+              record.absence_types = absenceTypes.map(function (sortedType) {
+                return record.absence_types.find(function (unsortedType) {
+                  return unsortedType && parseInt(unsortedType.id) === parseInt(sortedType.id);
+                });
+              });
+
+              return record;
+            });
+
+            ctrl.loadReportPage(page);
+            $rootScope.$digest();
+          });
+
+          it('sets .loading.report to false', function () {
+            expect(ctrl.loading.report).toBe(false);
+          });
+
+          it('sets .reportCount to the total number of records', function () {
+            expect(ctrl.reportCount).toBe(count);
+          });
+
+          it('sets .report to the response of AbsenceType.all call', function () {
+            expect(ctrl.report.length).toEqual(expected.length);
+          });
+
+          it('sorts .report[].absence_types in the same order as .absenceTypes', function () {
+            expect(ctrl.report).toEqual(expected);
+          });
+        });
+
+        describe('when there was an error loading the report`s page', function () {
+          beforeEach(function () {
+            leaveBalanceReport.all.and.returnValue($q.reject({
+              error_code: 'not-found',
+              error_message: 'Not Found.',
+              is_error: 1
+            }));
+
+            ctrl.loadReportPage(page);
+            $rootScope.$digest();
+          });
+
+          it('sets .loading.report to false', function () {
+            expect(ctrl.loading.report).toBe(false);
+          });
+
+          it('throws an error notification', function () {
+            expect(notification.error).toHaveBeenCalledWith(jasmine.any(String), jasmine.any(String));
```

Comment:
Not sure if we want to follow a specific error message pattern. To be honest, if we do need to follow one, I'd create a generic error handling service and pass the error to it.